plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,278 @@
+-- depending on whether 'BuiltinName' is saturated or not.
+
+evaluateFun :: EvaluationContext -> Heap -> Closure -> Closure-> LMachineResult
+evaluateFun ctx heap (Closure fun funEnv) argClosure =
+    case fun of
+      LamAbs _ var _ body ->
+          let (l, heap') = insertInHeap argClosure heap
+              env' = updateEnvironment (unUnique $ nameUnique var) l funEnv
+          in  computeL ctx heap' (Closure body env')
+
+      _ ->
+          -- Not a lambda: look for evaluation of a built-in function, possibly with some args already supplied.
+          -- We have to force the arguments, which means that we need to get the new heap back as well.
+          -- This bit is messy but should be much easier when we have n-ary application for builtins.
+          case computeL [] heap argClosure of
+            -- Force the argument, but only at the top level.
+            -- This is a bit of a hack.  We're not in the main compute/return process here.
+            -- We're preserving ctx, the context at entry, and we'll use that when we return.
+            -- We could also add a new type of stack frame for this.  Exactly what we do will
+            -- depend on the final interface to builtins.
+            Failure -> Failure
+            Success (Closure arg' env') heap' ->
+                let term = Apply () fun arg'
+                in case termAsPrimIterApp term of
+                     Nothing ->
+                         throw $ LMachineException "Trying to apply invalid term" term
+                         -- Was "Cannot reduce a not immediately reducible application."  This message isn't very helpful.
+                     Just (IterApp (StaticStagedBuiltinName name) spine) ->
+                         case applyEvaluateBuiltinName heap' env' name spine of
+                           ConstAppSuccess term' -> returnL ctx heap' (Closure term' funEnv)
+                           ConstAppStuck         -> returnL ctx heap' (Closure term  funEnv)
+                           -- It's arguable what the env should be here. That depends on what the built-in can return.
+                           -- Ideally it'd always return a closed term, so the environment should be irrelevant.
+                           ConstAppFailure       -> throw $ LMachineException "ConstAppFailure" term
+                           ConstAppError _err    -> throw $ LMachineException "ConstAppError" term
+
+                     Just (IterApp DynamicStagedBuiltinName{}     _    ) ->
+                         throw $ LMachineException "Dynamic builtins not supported yet" term
+
+
+-- | This is a workaround (thanks to Roman) to get things working while the dynamic builtins interface is under development.
+applyEvaluateBuiltinName :: Heap -> Environment -> BuiltinName -> [Value TyName Name ()] -> ConstAppResult
+applyEvaluateBuiltinName heap env name =
+    runIdentity . runEvaluate (const $ Identity . evalL heap env) . runQuoteT . applyBuiltinName name
```

Comment:
It should be changed latter, because the argument you ignore via `const` is the one that stores meaning of additional dynamic built-in names that may appear during the computation of a built-in application. You should take that argument and `mappend` it to the `DynamicBuiltinNameMeanings` map once you add that to the environment the machine runs in.